pkg/utils/utils_virtual.go

```diff
@@ -106,31 +110,65 @@ func metaData(platformType PlatformType, address string) (netFilter string, macA
 	return
 }
 
-func readOpenstackMetaData() (metaData *OSPMetaData, networkData *OSPNetworkData) {
-	networkData = &OSPNetworkData{}
-
-	rawBytes, err := ioutil.ReadFile(ospNetworkData)
+func getBodyFromUrl(url string) ([]byte, error) {
+	glog.V(2).Infof("Getting body from %s", url)
+	resp, err := retryablehttp.Get(url)
 	if err != nil {
-		glog.Errorf("error reading file %s, %v", ospNetworkData, err)
-		return
+		return nil, err
 	}
-
-	if err = json.Unmarshal(rawBytes, networkData); err != nil {
-		glog.Errorf("error unmarshalling raw bytes %v from %s", err, ospNetworkData)
-		return
+	rawBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
 	}
+	defer resp.Body.Close()
+	return rawBytes, nil
+}
 
+func unmarshalOpenStackMetaData(source string, rawBytes []byte) (metaData *OSPMetaData) {
 	metaData = &OSPMetaData{}
+	err := json.Unmarshal(rawBytes, metaData)
+	if err != nil {
+		glog.Errorf("error unmarshalling raw bytes %v from %s", err, source)
+	}
+	return
+}
 
-	rawBytes, err = ioutil.ReadFile(ospMetaData)
+func unmarshalOpenStackNetworkData(source string, rawBytes []byte) (networkData *OSPNetworkData) {
+	networkData = &OSPNetworkData{}
+	err := json.Unmarshal(rawBytes, networkData)
 	if err != nil {
-		glog.Errorf("error reading file %s, %v", ospMetaData, err)
-		return
+		glog.Errorf("error unmarshalling raw bytes %v from %s", err, source)
 	}
+	return
+}
 
-	if err = json.Unmarshal(rawBytes, metaData); err != nil {
-		glog.Errorf("error unmarshalling raw bytes %v from %s", err, ospNetworkData)
-		return
+func readOpenstackMetaData() (metaData *OSPMetaData, networkData *OSPNetworkData) {
+	// We first try to read metadata from config-drive and fallback to reading from
+	// OpenStack metadata service.
+	if rawBytes, err := ioutil.ReadFile(ospMetaDataFile); err == nil {
+		glog.V(2).Infof("getting OpenStack meta_data from config-drive")
+		unmarshalOpenStackMetaData(ospMetaDataFile, rawBytes)
+	} else {
+		glog.V(2).Infof("getting OpenStack meta_data from metadata server")
+		rawBytes, err := getBodyFromUrl(ospMetaDataUrl)
```

Comment:
I'm wondering if there would be a way to cache the data so it doesn't query the server every 10s when doing a reconcile. The metadata is not supposed to change unless the Machine had a change (e.g. via a new MachineConfig to add a new network interface).

What do you think and how could this be achieved?

Reply:
BTW the reconcile loop logs: https://paste.opendev.org/show/bNfTzI7Q79Vkkb10JUK2/

Reply:
ping @SchSeba ^

Reply:
We're calling this code once *per network device* every 30 seconds. We should definitely not do this.

I think it's appropriate to cache here. The calling code clearly assumes this is cheap, and we're changing that assumption. However, we should also move the call to metadata() out of the device loop in the parent because that's unnecessarily inefficient anyway. Ideally we'd explicitly populate this metadata from an initialiser once and cache it somewhere appropriate. The writer object?

Reply:
Do we expect this information to change? We certainly didn't expect it to change before as it was using config drive. If it will never change, there's no point in calling it every 30s. Even if it's just reading the config drive that's needlessly warming the planet.

Reply:
The information could change if the worker MachineSet would be updated to e.g. have one more NIC attached?

Reply:
Machines are explicitly immutable, so this would not happen via Machine updates. There could be some other mechanism to do it, though. Maybe CNI?

Reply:
Oh right. Then I don't see anything that could change, TBH.
I'll investigate Pierre's solution for now, which seems a great improvement.